### ISSUE TITLE
Partial Success with Status Details

### ISSUE THREAD

**Issue by qingling128** [google/google-api-ruby-client#544]
We are writing log entries to Logging API using the `write_entry_log_entries` [method](https://github.com/google/google-api-ruby-client/blob/master/generated/google/apis/logging_v2beta1/service.rb#L793), and we enabled `partial_success`.

When a request has a mix of failed and successful entries, it raised a `Google::Apis::ClientError` error. However, we can't find a way to get the status details to locate which entries are the failing ones.

To clarify it a bit, we are looking for something equivalent to the `WriteLogEntriesPartialErrors` mentioned in the documention of `partialSuccess` [here](https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/write).

Is this not supported yet?

**Comment by sqrrrl** [comment]
You should be able to get the full response body from the error -- the error has a `body` property that will be the raw body (as a string). You'll need to parse it yourself, but should be able to get the details you need from that.

**Comment by qingling128** [comment]
@sqrrrl Thanks for taking a look!

We recently fixed an issue for GCP instances. Previously when I tried on GCP, there is nothing in the error body. Now I'm able to extract some info as you recommended above. Yet I'm still having issues parsing the error body (Not necessarily due to this Ruby client lib. Could be the upper stream Logging API).

Basically the error body I got from the Ruby clients (both this one and the [grpc one](https://github.com/GoogleCloudPlatform/google-cloud-ruby/issues/1239)) has weird format (I added some new lines to make it readable)

```
{  
	"error": {    
		"code": 400,    
		"message": "Resource type cannot be empty",    
		"errors": [      
			{        
				"message": "Resource type cannot be empty",        
				"domain": "global",        
				"reason": "badRequest",        
				"debugInfo": "
					detail: \\"[ORIGINAL ERROR] generic::invalid_argument: Resource type cannot be empty [google.rpc.error_details_ext] {
					message: \\\\\\"Resource type cannot be empty\\\\\\" 
					details { 
						type_url: \\\\\\"type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors\\\\\\" 
						value: \\\\\\"\\\\\\\%\\\\\\\\010\\\\\\\\000\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\\%\\\\\\\\010\\\\\\\\001\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\\%\\\\\\\\010\\\\\\\\002\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\" 
					} 
				}\\"\"      
			}    
		],    
		"status": "INVALID_ARGUMENT"  
	}
}
```

However when I call the Logging API directly, I got:

```
{
  "error": {
    "code": 400,
    "message": "Illegal character ':' in log name",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors",
        "logEntryErrors": {
          "0": {
            "code": 3,
            "message": "Illegal character ':' in log name"
          },
          "1": {
            "code": 7,
            "message": "User not authorized."
          }
        }
      },
      {
        "@type": "type.googleapis.com/google.rpc.DebugInfo",
        "detail": "[ORIGINAL ERROR] generic::invalid_argument: Illegal character ':' in log name [google.rpc.error_details_ext] { 
message: \"Illegal character \\':\\' in log name\" details { 
type_url: \"type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors\" 
value: \"\\n)\\010\\000\\022%\\010\\003\\022!Illegal character \\':\\' in log name\\n
\\034\\010\\001\\022\\030\\010\\007\\022\\024User not authorized.\" } }"
      }
    ]
  }
}
```

Note that "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors" seems to be what we want, And I would be fine with "type.googleapis.com/google.rpc.DebugInfo" if I could successfully parse out the indexes (right now it seems kind of messed up).

I'm still working with the Logging API team to try to figure out what the expected behavior and fix should be.

My hunch is we either somehow fix "type.googleapis.com/google.rpc.DebugInfo" or use "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors" instead if possible.

Will keep you updated.

**Comment by qingling128** [comment]
Hi @sqrrrl, after some investigation with Logging API team it turned out that I got weird error format like

```
    details { 
        type_url: \\\\\\"type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors\\\\\\" 
	value: \\\\\\"\\\\\\\%\\\\\\\\010\\\\\\\\000\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\\%\\\\\\\\010\\\\\\\\001\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\\%\\\\\\\\010\\\\\\\\002\\\\\\\\022!\\\\\\\\010\\\\\\\\003\\\\\\\\022\\\\\\\\035Resource type cannot be empty\\\\\\" 
    } 
```

because this is the V1 format. The V2 format would have the error details in well-formed json:

```
      {
        "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors",
        "logEntryErrors": {
          "0": {
            "code": 3,
            "message": "Illegal character ':' in log name"
          },
          "1": {
            "code": 7,
            "message": "User not authorized."
          }
        }
      },
```

To force v2 format, the client should use either the query param: `$apiFormat=2` or the header: `X-GOOG-API-FORMAT-VERSION: 2`. Can we force this client to use V2 format?

**Comment by sqrrrl** [comment]
Can't force it without breaking existing clients. Can make it opt-in though if you need it.

I had wanted to move to v2 a while back, but there were a few problems:

* Older APIs only support v1. Asking for v2 still gives v1 formats.
* There's no signal from the API definitions that a given API supports V2
* There's nothing in the response that indicates which format an error is. Have to infer from the content.
* Doing so is fragile -- if, for example, the older backends started supporting v2 errors it would manifest as a visible breaking changing in the client since it was asking for v2 by default.

Safest thing was to just force v1 for everything.

LMK if you need it and I'll see about adding an opt-in approach + parsing the errors correctly.

**Comment by qingling128** [comment]
Opt-in sounds good to me. The v1 format makes it impossible to extract the index of the failing entries, which is important to make partial success very useful. Can we default to v1 as is and support v2 as opt-in?

**Comment by qingling128** [comment]
Hi, any news on this one?

**Comment by qingling128** [comment]
Verified the fix with `google-api-client (0.11.3)`. Closing this issue. Thanks Steve!

```
$ cat partial_error_non_grpc.rb
require 'google/apis/logging_v2'
require 'googleauth'

client = Google::Apis::LoggingV2::LoggingService.new
client.authorization = Google::Auth.get_application_default(
  'https://www.googleapis.com/auth/logging.write')

resource = Google::Apis::LoggingV2::MonitoredResource.new(
  type: 'gce_instance',
  labels: {}
)

entries = [
  Google::Apis::LoggingV2::LogEntry.new(
    {
      resource: resource,
      log_name: 'projects/gapic-test/logs/my-log',
      text_payload: 'pass'
    }
  ),
  Google::Apis::LoggingV2::LogEntry.new(
    {
      resource: resource,
      log_name: 'projects/gapic-test/logs/bad:log',
      text_payload: 'fail'
    }
  )
]

write_request = \
  Google::Apis::LoggingV2::WriteLogEntriesRequest.new(
    log_name: 'projects/gapic-test/logs/my-log',
    resource: resource,
    labels: {},
    partial_success: true,
    entries: entries)

begin
  client.write_entry_log_entries(
    write_request,
    options: {
      api_format_version: '2'
    }
  )
rescue => error
  puts "error:\n#{error.inspect}\n"
  puts "error body =\n#{error.body}\n"
end

$ ruby partial_error_non_grpc.rb
error:
#<Google::Apis::ClientError: PERMISSION_DENIED: User not authorized.>
error body =
{
  "error": {
    "code": 403,
    "message": "User not authorized.",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors",
        "logEntryErrors": {
          "0": {
            "code": 7,
            "message": "User not authorized."
          },
          "1": {
            "code": 3,
            "message": "Log name contains illegal character :"
          }
        }
      },
      {
        "@type": "type.googleapis.com/google.rpc.DebugInfo",
        "detail": "[ORIGINAL ERROR] generic::permission_denied: User not authorized. [google.rpc.error_details_ext] { message: \"User not authorized.\" details { type_url: \"type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors\" value: \"\\n\\034\\010\\000\\022\\030\\010\\007\\022\\024User not authorized.\\n-\\010\\001\\022)\\010\\003\\022%Log name contains illegal character :\" } }"
      }
    ]
  }
}

$ gem list | grep google-api-client
google-api-client (0.11.3)
```
